torch/distributed/elastic/agent/server/api.py

```diff
@@ -60,6 +61,8 @@ class WorkerSpec:
                      if not specified then will chose a random free port
         master_addr: fixed master_addr to run the c10d store on rank 0
                      if not specified then will chose hostname on agent rank 0
+        endpoint: original endpoint that is discarded when the static
```

Comment:
Also commented on https://github.com/pytorch/pytorch/issues/73656

Reply:
I think the main problem is that PyTorch tries to be too smart here, so whatever the user passes should not just be discarded. :)

It seems to me like the issue lies in [`_get_addr_and_port`](https://github.com/pytorch/pytorch/blob/cda8635a5e9822f08440f4ef89f7e70a6977b1fe/torch/distributed/launcher/api.py#L157-L161), where the discarding happens. If [`SimpleElasticAgent._set_master_addr_port`](https://github.com/pytorch/pytorch/blob/cda8635a5e9822f08440f4ef89f7e70a6977b1fe/torch/distributed/elastic/agent/server/api.py#L506-L515) were to use the same information as [`_matches_machine_hostname`](https://github.com/pytorch/pytorch/blob/cda8635a5e9822f08440f4ef89f7e70a6977b1fe/torch/distributed/elastic/rendezvous/c10d_rendezvous_backend.py#L133-L144) – meaning that address and port are parsed from the endpoint when the backend is not `"static"` (also meaning that they are never `None` anymore) – that would probably be another solution (instead of adding a `WorkerSpec.endpoint`).

Then we would
- set `spec.master_{addr,port}` to the parsed values,
- remove the `Optional`ity of their type, and
- we would not need to add a new `endpoint` parameter since all required information is in the `WorkerSpec`.

While this would probably be valid, there are issues with this approach as you mentioned:

> the fact that we either take master_addr and master_port (if rdzv_backend="static") OR we split rdzv_endpoint by : and pass it in as master_addr and master_port is intentional. (--master_addr, and --master_port) and --rdzv_endpoint are mutually exclusive. It would be confusing to make them complimentary.

Just like you said here, I didn't want to make the two complimentary because it seemed too disruptive to whatever the code expects.

Reply:
I see the issue you were having. Take a look at my comment on the issue: https://github.com/pytorch/pytorch/issues/73656#issuecomment-1204924739

I think the correct way to fix this is to add a if-statement in https://github.com/pytorch/pytorch/blob/cda8635a5e9822f08440f4ef89f7e70a6977b1fe/torch/distributed/launcher/api.py#L157 to handle the case where `rdzv_backend == "static"` (and `is_host` is specified in the rdzv config) then "reuse" the master (port still needs to be left as `None` so that a free random port is assigned for the worker MASTER_PORT that will be used by torch distributed to initialize the process group on).

As you may have already noticed, there are two TCPStores when the `c10d` rdzv backend is used:

1) At the node-level (on the agent rank 0) to coordinate rendezvous (specified by `--rdzv_endpoint`)
2) At the trainer (worker) level (on rank 0) to initialize the pytorch process group (elastic currently only lets you override this when you use `--rdzv_backend=static`)